From ff75ea02b109da48e4ed5cbef3f62a71a3c0c263 Mon Sep 17 00:00:00 2001
From: Xie XiuQi <xiexiuqi@huawei.com>
Date: Fri, 26 Feb 2021 20:20:59 +0800
Subject: [PATCH] arm64/mpam: add mpam extension runtime detection
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: ff75ea02b109da48e4ed5cbef3f62a71a3c0c263
Modified-by-SEL: Yes, modify ARM64_HAS_MPAM definition to avoid conflict with 5.10.105

hulk inclusion
category: feature
feature: ARM MPAM support
bugzilla: 48265
CVE: NA

--------------------------------

According to <<MPAM Extension Architecture>> document, we add MPAM
extension feature runtime detection.

Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Wang ShaoBo <bobo.shaobowang@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
Signed-off-by: Chenxi Mao <chenxi.mao@suse.com>
---
 arch/arm64/include/asm/cpucaps.h |    1 +
 arch/arm64/include/asm/sysreg.h  |    1 +
 arch/arm64/kernel/cpufeature.c   |   12 ++++++++++++
 arch/arm64/kernel/mpam.c         |    3 +++
 4 files changed, 17 insertions(+)

--- a/arch/arm64/include/asm/cpucaps.h
+++ b/arch/arm64/include/asm/cpucaps.h
@@ -67,6 +67,7 @@
 #define ARM64_SPECTRE_BHB			59
 #define ARM64_HAS_ECV				64
 #define ARM64_HAS_EPAN				65
+#define ARM64_HAS_MPAM				66
 #define ARM64_NCAPS				80
 
 #endif /* __ASM_CPUCAPS_H */
--- a/arch/arm64/include/asm/sysreg.h
+++ b/arch/arm64/include/asm/sysreg.h
@@ -738,6 +738,7 @@
 #define ID_AA64PFR0_EL1_SHIFT		4
 #define ID_AA64PFR0_EL0_SHIFT		0
 
+#define ID_AA64PFR0_MPAM		0x1
 #define ID_AA64PFR0_AMU			0x1
 #define ID_AA64PFR0_SVE			0x1
 #define ID_AA64PFR0_RAS_V1		0x1
--- a/arch/arm64/kernel/cpufeature.c
+++ b/arch/arm64/kernel/cpufeature.c
@@ -1913,6 +1913,18 @@ static const struct arm64_cpu_capabiliti
 		.cpu_enable = cpu_clear_disr,
 	},
 #endif /* CONFIG_ARM64_RAS_EXTN */
+#ifdef CONFIG_MPAM
+	{
+		.desc = "ARM64 MPAM Extension Support",
+		.capability = ARM64_HAS_MPAM,
+		.type = ARM64_CPUCAP_SCOPE_SYSTEM,
+		.matches = has_cpuid_feature,
+		.sys_reg = SYS_ID_AA64PFR0_EL1,
+		.sign = FTR_UNSIGNED,
+		.field_pos = ID_AA64PFR0_MPAM_SHIFT,
+		.min_field_value = ID_AA64PFR0_MPAM,
+	},
+#endif /* CONFIG_MPAM */
 #ifdef CONFIG_ARM64_AMU_EXTN
 	{
 		/*
--- a/arch/arm64/kernel/mpam.c
+++ b/arch/arm64/kernel/mpam.c
@@ -406,6 +406,9 @@ static __init bool get_rdt_mon_resources
 
 static __init bool get_resctrl_resources(void)
 {
+	if (!cpus_have_const_cap(ARM64_HAS_MPAM))
+		return false;
+
 	rdt_alloc_capable = get_rdt_alloc_resources();
 	rdt_mon_capable = get_rdt_mon_resources();
 
